resources/aws_ec2.go

```diff
@@ -1014,6 +994,78 @@ func (obj *AwsEc2Res) prependName() string {
 	return AwsPrefix + obj.GetName()
 }
 
+// waitForStopAndTerminate waits for the instance to stop and waits for it to
+// terminate. If either waiter returns, the instance state is checked, and an
+// awsEc2Event is returned.
+func (obj *AwsEc2Res) waitForStopAndTerminate(ctx context.Context, waitInput *ec2.DescribeInstancesInput) (awsEc2Event, error) {
+	if err := obj.waitUntilInstanceStoppedOrTerminatedWithContext(ctx, waitInput); err != nil {
+		return awsEc2EventNone, errwrap.Wrapf(err, "error waiting for instance to stop or terminate")
+	}
+	// Check the instance state, and return the appropriate event.
+	stateOutput, err := obj.client.DescribeInstances(waitInput)
+	if err != nil {
+		return awsEc2EventNone, errwrap.Wrapf(err, "error describing instances")
+	}
+	if len(stateOutput.Reservations) == 1 {
+		switch *stateOutput.Reservations[0].Instances[0].State.Name {
+		case "stopped":
+			return awsEc2EventInstanceStopped, nil
+		case "terminated":
+			return awsEc2EventInstanceTerminated, nil
+		}
+	}
+	return awsEc2EventNone, nil
+}
+
+func (obj *AwsEc2Res) waitUntilInstanceStoppedOrTerminatedWithContext(ctx context.Context, waitInput *ec2.DescribeInstancesInput) error {
+	errChan := make(chan error)
+	defer close(errChan)
+	wg := sync.WaitGroup{}
+	defer wg.Wait()
+	wgCtx, wgCancel := context.WithCancel(ctx)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer wgCancel()
+		if err := obj.client.WaitUntilInstanceStoppedWithContext(wgCtx, waitInput); err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				if aerr.Code() == request.CanceledErrorCode || aerr.Code() == request.WaiterResourceNotReadyErrorCode {
+					return
+				}
+			}
+			select {
+			case errChan <- errwrap.Wrapf(err, "unknown error waiting for instance to stop"):
+			case <-obj.closeChan:
+			}
+		}
+	}()
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		defer wgCancel()
+		if err := obj.client.WaitUntilInstanceTerminatedWithContext(wgCtx, waitInput); err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				if aerr.Code() == request.CanceledErrorCode || aerr.Code() == request.WaiterResourceNotReadyErrorCode {
+					return
+				}
+			}
+			select {
+			case errChan <- errwrap.Wrapf(err, "unknown error waiting for instance to terminate"):
+			case <-obj.closeChan:
+			}
+		}
+	}()
+	select {
```

Comment:
This pattern isn't correct-- the two go routines will start running, and then assuming something isn't immediately available on errChan, default will trigger mean we *will* return nil. The only catch is we'll only see that happen after the two go-routines close and the waitGroup lets us exit.

So you've actually just masked any chance of error.